feat: adds tf-terminology-breakdown article (#49)
masterpointio/masterpoint.io: content/updates/tf-terminology-breakdown.md

---
visible: true
draft: false
title: Terraform + OpenTofu Terminology Breakdown
slug: terraform-opentofu-terminology-breakdown
author: Matt Gowie
date: 2025-03-01
description: List of Terraform and OpenTofu terms with definitions and explanations.

> image: /img/updates/terraform-null-label-part1.png

Author: @oycyc mind helping me come up with an image for this one? You've done some good images in the past and I think I'm largely bad at prompting those. Interested in hearing your thought process around that too. The images I tried for this were 👎

Reviewer: @Gowiem you want something abstract like this?
![cloud](https://github.com/user-attachments/assets/e3c23145-4443-4b87-9f82-7bb888642e6a)

Reviewer 2: Normally I think I would've done an abstract image like the above then added some text and icons related to the stuff in the article, but just tried OpenAI's new image generator (https://openai.com/index/introducing-4o-image-generation/) and it's kinda neat, this seems like it's gonna take all graphic designers jobs away 😓 :
![image](https://github.com/user-attachments/assets/78b87ec1-8232-401a-ba25-0e0d74e95f78)

callout: <p>👋 <b>Got a term that you're still confused on in the TF or IaC space that you want us to include here? <a href='/contact'>Get in touch and we'd be happy to add it!</a>
---

The Terraform and OpenTofu (collectively referred to as "TF") ecosystem has evolved significantly since HashiCorp's initial release, and the growing sea of tooling and management platforms has led to some interesting (and occasionally frustrating) terminology fragmentation. We're going to break down some essential TF terminology that you'll encounter: workspaces, modules, state management, testing frameworks, and more.

## Workspaces

Let's start with one of the most commonly misunderstood concepts: workspaces. The term "workspace" has different meanings depending on whether you're using Terraform CLI or Terraform Cloud, so it's important to be aware of the context you're working in when using them.

> ### TF CLI Workspaces

Reviewer 2: Wonder what just one # will look like?

Because ## vs ### is not too strong of a difference on the website and gets jumbled together whereas this is supposed to show that TF CLI Workspaces is a type of Workspace

Author: @oycyc I tried this out... the H1 headers are probably too big? We may want to revert that change and then work on fixing the sizing between H2 + H3

Reviewer: Yeah, I agree that H1 is too big. It would be nice to fix the sizing diff anyway 👍

Here is an example of a child module: [cloudposse/terraform-aws-vpc](https://github.com/cloudposse/terraform-aws-vpc)

## Submodules

While often used interchangeably with "child modules", submodules typically refer to modules nested within another module's directory structure. These aren't modules defined within the same file as their parent - they're separate module directories that physically exist inside the parent module's directory.

For example, you might have a directory structure like this:

```txt
vpc/
├── main.tf
├── variables.tf
├── outputs.tf
└── modules/
  └── vpc-endpoints/
      ├── main.tf
      ├── variables.tf
      └── outputs.tf
```

In this case, "vpc-endpoints" is a submodule of the "vpc" parent. This nesting capability provides additional flexibility in organizing your infrastructure code, but it's worth noting that if you're building many submodules within a child module, it may just be time to break up your child module into separate, stand-alone modules.

Here is an example of a submodule: [cloudposse/terraform-aws-vpc/tree/main/modules/vpc-endpoints](https://github.com/cloudposse/terraform-aws-vpc/tree/main/modules/vpc-endpoints)

> # TF Organization Patterns

Reviewer: I'd move this to the end. It's our internal term, which I believe should be shared and popularized, but I don't want to create an impression (almost at the beginning of the article) that we're introducing something new here. The focus is on widely used terms and possible related confusion.

Author: Updated!

# Industry Terms

As the Terraform ecosystem has matured, new terms have emerged beyond the core concepts. These terms represent specialized tools, architectural patterns, and platform-specific nomenclature that you'll encounter when working with Terraform at scale.

## Provider

A provider in TF is a plugin that enables interaction with an API; typically for a specific infrastructure platform or service. Providers serve as the bridge between TF configuration code and the actual infrastructure resources you're managing. Each provider (like AWS, Azure, GCP, Kubernetes, GitHub, CloudFlare, DataDog, etc.) contains resource types and data sources that correspond to services offered by that platform. Providers must be configured in your TF code with the necessary authentication credentials and regional settings before you can use their associated resources.

The term "Provider" has also been adopted by the wider IaC tools like Crossplane + Pulumi.

## TACOS

> TACOS stands for "Terraform Automation and Collaboration Software." This term is associated with the various Terraform automation products in the space like HCP Terraform, Spacelift, Env0, Scalr, and similar. These are generally hosted platforms that serve as execution environments for TF that help teams collaborate on infrastructure changes, manage workflows, handle approvals, and provide visibility into the state of the organization's infrastructure across multiple environments and projects.

Reviewer: ```suggestion
TACOS stands for "Terraform Automation and Collaboration Software." This term is associated with the various Terraform automation products in the space like [HCP Terraform](https://cloud.hashicorp.com/products/terraform), [Spacelift](https://spacelift.io/), [Env0](https://www.env0.com/), [Scalr](https://www.scalr.com/), and similar. These are generally hosted platforms that serve as execution environments for TF that help teams collaborate on infrastructure changes, manage workflows, handle approvals, and provide visibility into the state of the organization's infrastructure across multiple environments and projects.
```

## TF Frameworks

TF frameworks are wrapper projects that help you with some of the tedium in Terraform and OpenTofu. Frameworks tend to be lighter-weight implementations that typically wrap CLI execution

These tools execute TF on your behalf and as a result they provide benefits like convention focused directory structures, dynamic backend configuration, root module dependencies, scripting hooks, and similar that Vanilla TF does not. The primary players in this space are [Terragrunt](https://terragrunt.gruntwork.io/), [Terramate](https://terramate.io/), [Atmos](https://atmos.tools/), and [Terraspace](https://terraspace.cloud/).

At Masterpoint, we've used ALL of these frameworks through either greenfield projects or through helping clients dig themselves out of an existing hole that they ended up in. The choice of whether to use a framework or not could take up a full post, so we'll save our thoughts on this topic for later.

## Stack

> The term "stack" has one of the most ambiguous definitions. Different platforms in the Terraform ecosystem refer to different concepts when they use this term. Here's how different platforms define "stacks":

Reviewer: ```suggestion
The term "stack" has one of the most ambiguous definitions. Different platforms in the TF ecosystem refer to different concepts when they use this term. Here's how different platforms define "stacks":
```

> * Spacelift: a deployment unit with its own state and configuration: [https://docs.spacelift.io/concepts/stack](https://docs.spacelift.io/concepts/stack)
> * Atmos: a collection of components and modules: [https://docs.cloudposse.com/resources/legacy/stacks/](https://docs.cloudposse.com/resources/legacy/stacks/)
> * HashiCorp: TF configurations organized into components across multiple environments [https://developer.hashicorp.com/terraform/language/stacks](https://developer.hashicorp.com/terraform/language/stacks)
> * Terramate: Collections of resources managed as a unit [https://terramate.io/docs/cli/stacks/](https://terramate.io/docs/cli/stacks/)
> * [Te](https://terramate.io/docs/cli/stacks/)rragrunt: A collection of "units" (single instance of infrastructure) managed by Terragrunt: [https://terragrunt.gruntwork.io/docs/getting-started/terminology/#stack](https://terragrunt.gruntwork.io/docs/getting-started/terminology/#stack)

Reviewer: ```suggestion
* Spacelift: A deployment unit with its own state and configuration, see [Main Concepts - Stack](https://docs.spacelift.io/concepts/stack).
* Atmos: A collection of components and modules, see [Resources - Stacks](https://docs.cloudposse.com/resources/legacy/stacks/).
* HashiCorp: TF configurations organized into components across multiple environments, see [Stacks overview](https://developer.hashicorp.com/terraform/language/stacks).
* Terramate: Collections of resources managed as a unit, see [Stacks](https://terramate.io/docs/cli/stacks/).
* Terragrunt: A collection of "units" (a single instance of infrastructure) managed by Terragrunt, see [Terminology - Stack](https://terragrunt.gruntwork.io/docs/getting-started/terminology/#stack).
```

Platforms like Env0 and Scalr, which provide a more holistic, top-down management of infrastructure configurations like Terraform and Kubernetes, wrap Terragrunt's toolchain to provide stack functionality.

## Terralith Pattern

The [Terralith pattern](https://masterpoint.io/updates/terralith-monolithic-terraform-architecture/) represents a specific approach to TF architecture that opts for a single state file for all infrastructure using a monolithic root module approach. While it's often seen as a natural starting point for new projects, teams should view it as technical debt that needs to be addressed in the future. The pattern's fundamental limitations - from state file locking issues to increased blast radius of changes - make it unsuitable for long-term use or any environment that needs to scale beyond basic infrastructure management.

> Dealing with the Terralith problem and want to know more? [Check out our original post on this topic here](https://masterpoint.io/updates/terralith-monolithic-terraform-architecture/) and then [learn how to break up a Terralith in our follow up article here](https://masterpoint.io/updates/steps-to-break-up-a-terralith/).

Reviewer: ```suggestion
Dealing with the Terralith problem and want to know more? [Check out our original post on this topic here](https://masterpoint.io/updates/terralith-monolithic-terraform-architecture/) and then [learn how to break up a Terralith in our follow-up article here](https://masterpoint.io/updates/steps-to-break-up-a-terralith/).
```

---
visible: true
draft: false
title: Terraform + OpenTofu Terminology Breakdown
slug: terraform-opentofu-terminology-breakdown
author: Matt Gowie
date: 2025-03-01
description: List of Terraform and OpenTofu terms with definitions and explanations.
image: /img/updates/terraform-null-label-part1.png
callout: <p>👋 <b>Got a term that you're still confused on in the TF or IaC space that you want us to include here? <a href='/contact'>Get in touch and we'd be happy to add it!</a>
---

# Table of Contents

> - [Table of Contents](#table-of-contents)

Reviewer 3: I'd vote for removing this line - the reader knows there's a table of contents because that's where they're reading

Author: Ha this was actually a pain because the tool I was using kept adding it back. Done in latest updates though.

  - [Stack](#stack)
  - [Terralith Pattern](#terralith-pattern)
  - [Demystifying HashiCorp Offerings: Terraform Cloud vs Terraform Enterprise vs HCP Terraform](#demystifying-hashicorp-offerings-terraform-cloud-vs-terraform-enterprise-vs-hcp-terraform)
- [TF Organization Patterns](#tf-organization-patterns)
  - [Multi-instance Root Modules](#multi-instance-root-modules)
  - [Single-instance Root Modules](#single-instance-root-modules)
- [Wrapping Up](#wrapping-up)

# Intro

The Terraform and OpenTofu (collectively referred to as "TF") ecosystem has evolved significantly since HashiCorp's initial release, and the growing sea of tooling and management platforms has led to some interesting (and occasionally frustrating) terminology fragmentation. We're going to break down some essential TF terminology that you'll encounter: workspaces, modules, state management, testing frameworks, and more.

# Workspaces

Let's start with one of the most commonly misunderstood concepts: workspaces. The term "workspace" has different meanings depending on whether you're using Terraform CLI or Terraform Cloud, so it's important to be aware of the context you're working in when using them.

> ## TF CLI Workspaces

Author: Not a bad idea, but I think we leave this off for now.

In the context of TF CLI, workspaces are named, isolated instances of state within a single backend. Think of them as separate environments for managing different versions of the same infrastructure. A common use case is maintaining development, staging, and production infrastructure environments using the same TF configuration.

> Here's where people often get tripped up: CLI workspaces are not equivalent to long-lived branches in your version control system. Unlike branches, you don't "promote" or merge changes from one workspace to another. Each workspace maintains its own state file, but they all use the same configuration code. The differences between workspaces typically come from the variable values you apply; for example, setting different instance sizes or replica counts based on environment needs.

Reviewer 3: > Each workspace maintains its own state file,

I think this is an important boundary and logical way to think about tf workspaces. Maybe add a call out or quote to highlight this point.
